tailnet/service.go

```diff
@@ -230,3 +238,85 @@ func (c communicator) loopResp() {
 		}
 	}
 }
+
+type NetworkTelemetryBatcher struct {
+	clock     clock.Clock
+	frequency time.Duration
+	maxSize   int
+	batchFn   func(batch []*proto.TelemetryEvent)
+
+	mu      sync.Mutex
+	closed  chan struct{}
+	done    chan struct{}
+	ticker  *clock.Ticker
+	pending []*proto.TelemetryEvent
+}
+
+func NewNetworkTelemetryBatcher(clk clock.Clock, frequency time.Duration, maxSize int, batchFn func(batch []*proto.TelemetryEvent)) *NetworkTelemetryBatcher {
+	b := &NetworkTelemetryBatcher{
+		clock:     clk,
+		frequency: frequency,
+		maxSize:   maxSize,
+		batchFn:   batchFn,
+		closed:    make(chan struct{}),
+		done:      make(chan struct{}),
+	}
+	b.start()
+	return b
+}
+
+func (b *NetworkTelemetryBatcher) Close() error {
+	close(b.closed)
+	<-b.done
+	return nil
+}
+
+func (b *NetworkTelemetryBatcher) sendTelemetryBatch() {
+	b.mu.Lock()
+	defer b.mu.Unlock()
+	events := b.pending
+	b.pending = []*proto.TelemetryEvent{}
+	go b.batchFn(events)
+}
+
+func (b *NetworkTelemetryBatcher) start() {
+	b.mu.Lock()
+	defer b.mu.Unlock()
+	ticker := b.clock.NewTicker(b.frequency)
+	b.ticker = ticker
+
+	go func() {
+		defer close(b.done)
+		defer ticker.Stop()
+
+		for {
+			select {
+			case <-ticker.C:
+				b.sendTelemetryBatch()
+			case <-b.closed:
+				// Send any remaining telemetry events before exiting.
+				b.sendTelemetryBatch()
+				return
+			}
+		}
+	}()
+}
+
+func (b *NetworkTelemetryBatcher) Handler(events []*proto.TelemetryEvent) {
+	b.mu.Lock()
+	defer b.mu.Unlock()
+
+	for _, event := range events {
+		b.pending = append(b.pending, event)
+
+		if len(b.pending) >= b.maxSize {
+			events := b.pending
+			b.pending = []*proto.TelemetryEvent{}
+			// Perform the send in a goroutine to avoid blocking the DRPC call.
+			if b.ticker != nil {
+				b.ticker.Reset(b.frequency)
```

Comment:
The ticker could be stopped when you reset it, if Handler is called after Close.